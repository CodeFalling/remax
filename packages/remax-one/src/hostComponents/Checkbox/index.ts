import createHostComponent from '../../createHostComponent';

export interface CheckboxProps extends React.AriaAttributes {
  // 通用属性
  readonly dataset?: DOMStringMap;
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  /** checkbox标识，选中时触发checkbox-group的 change 事件，并携带 checkbox 的 value */
  value?: string;
  /** 当前是否选中，可用来设置默认选中 */
  checked?: boolean;
  /** 是否禁用 */
  disabled?: boolean;
  /** checkbox的颜色，同css的color */
  color?: string;
}

export default createHostComponent<CheckboxProps>('checkbox');
