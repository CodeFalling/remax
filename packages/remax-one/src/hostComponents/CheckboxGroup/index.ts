import createHostComponent from '../../createHostComponent';
import { InputEvent } from '../../types';

export interface CheckboxGroupProps {
  readonly dataset?: DOMStringMap;
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  name?: string;
  onChange?: (e: InputEvent) => void;
}

export default createHostComponent<CheckboxGroupProps>('checkbox-group');
