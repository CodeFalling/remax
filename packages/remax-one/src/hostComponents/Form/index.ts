import * as React from 'react';
import createHostComponent from '../../createHostComponent';
import { FormEvent } from '../../types';

export interface FormProps {
  // 通用属性
  readonly dataset?: DOMStringMap;
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  onSubmit?: (e: FormEvent) => void;
  onReset?: (e: FormEvent) => void;
}
const Form = createHostComponent<FormProps>('form');

export default Form;
