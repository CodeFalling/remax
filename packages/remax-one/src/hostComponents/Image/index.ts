import createHostComponent from '../../createHostComponent';
import ImageProps from './props';

const Image = createHostComponent<ImageProps>('image');

export default Image;
