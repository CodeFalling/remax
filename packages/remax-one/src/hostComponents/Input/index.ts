import * as React from 'react';
import { InputEvent } from '../../types';
import { createCallback, createInputEvent } from '../../createHostComponent';

export interface InputProps {
  // 通用属性
  readonly dataset?: DOMStringMap;
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  /** 输入框的初始内容 */
  defaultValue?: string;
  value?: string;
  name?: string;
  /** input 的类型 */
  type?: string;
  /** 是否是密码类型 */
  password?: boolean;
  /** 输入框为空时占位符 */
  placeholder?: string;
  placeholderStyle?: React.CSSProperties;
  /** 是否禁用 */
  disabled?: boolean;
  /** 最大输入长度，设置为 -1 的时候不限制最大长度 */
  maxlength?: number;
  /** 获取焦点 */
  focus?: boolean;
  onInput?: (e: InputEvent) => any;
  onConfirm?: (e: InputEvent) => void;
  onFocus?: (e: InputEvent) => void;
  onBlur?: (e: InputEvent) => void;
}

export interface InputState {
  value?: string;
  controlled: boolean;
}

export default class Input extends React.Component<InputProps, InputState> {
  static defaultProps = {
    'toutiao-maxlength': 140,
    'toutiao-selection-end': 999,
    'toutiao-selection-start': 999,
    'wechat-maxlength': 140,
    'wechat-selection-end': 999,
    'wechat-selection-start': 999,
  };

  state: InputState = {
    value: '',
    // 支付宝
    controlled: false,
  };

  constructor(props: InputProps) {
    super(props);

    const controlled = props.value !== undefined;
    const value = controlled ? props.value : props.defaultValue;

    this.state = {
      value,
      controlled,
    };
  }

  componentDidUpdate() {
    if (this.props.value !== undefined && this.props.value !== this.state.value) {
      this.setState({ value: this.props.value });
    }
  }

  handleInput = (e: InputEvent) => {
    const { controlled } = this.state;

    if (!controlled) {
      this.setState({ value: e.target.value });
    }

    if (typeof this.props.onInput === 'function') {
      return this.props.onInput(e);
    }

    // 微信
    // 注意，微信要对 input 受控，必须要在 onInput 方法返回值
    return controlled ? this.props.value : e.target.value;
  };

  render() {
    const inputProps = {
      ...this.props,
      onInput: createCallback(this.handleInput, createInputEvent),
    };

    if (inputProps.onConfirm) {
      inputProps.onConfirm = createCallback(this.props.onConfirm, createInputEvent);
    }

    if (inputProps.onFocus) {
      inputProps.onFocus = createCallback(this.props.onFocus, createInputEvent);
    }

    if (inputProps.onBlur) {
      inputProps.onBlur = createCallback(this.props.onBlur, createInputEvent);
    }

    return React.createElement('input', {
      ...inputProps,
      ...this.state,
    });
  }
}
