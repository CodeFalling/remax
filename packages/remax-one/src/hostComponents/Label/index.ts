import createHostComponent from '../../createHostComponent';
import { LabelProps } from './props';

export default createHostComponent<LabelProps>('label');
