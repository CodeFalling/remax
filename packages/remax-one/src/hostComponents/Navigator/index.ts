import createHostComponent from '../../createHostComponent';
import { NavigatorProps } from './props';

export default createHostComponent<NavigatorProps>('navigator');
