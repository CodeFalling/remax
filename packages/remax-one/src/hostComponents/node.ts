import * as button from './Button/node';
import * as form from './Form/node';
import * as image from './Image/node';
import * as input from './Input/node';
import * as label from './Label/node';
import * as radioGroup from './RadioGroup/node';
import * as radio from './Radio/node';
import * as text from './Text/node';
import * as textarea from './Textarea/node';
import * as view from './View/node';
import * as webView from './WebView/node';

const hostComponents = new Map();

hostComponents.set('button', button);
hostComponents.set('form', form);
hostComponents.set('image', image);
hostComponents.set('input', input);
hostComponents.set('label', label);
hostComponents.set('radio-group', radioGroup);
hostComponents.set('radio', radio);
hostComponents.set('text', text);
hostComponents.set('textarea', textarea);
hostComponents.set('view', view);
hostComponents.set('web-view', webView);

export default hostComponents;
