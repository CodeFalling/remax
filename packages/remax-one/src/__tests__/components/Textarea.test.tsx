import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import { Textarea } from '../../hostComponents';

describe('Textarea', () => {
  it('render correctly', () => {
    const testRenderer = TestRenderer.create(<Textarea className="class" />);

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('default value', () => {
    const testRenderer = TestRenderer.create(<Textarea className="class" defaultValue="1" />);

    const instance = testRenderer.root.findByType('textarea');

    expect(instance.props.value).toEqual('1');

    instance.props.onInput({
      target: {},
      currentTarget: {},
      detail: {
        value: '2',
      },
    });

    expect(instance.props.value).toEqual('2');
  });

  it('value', () => {
    const testRenderer = TestRenderer.create(<Textarea className="class" value="1" />);

    const instance = testRenderer.root.findByType('textarea');

    expect(instance.props.value).toEqual('1');

    instance.props.onInput({
      target: {},
      currentTarget: {},
      detail: {
        value: '2',
      },
    });

    expect(instance.props.value).toEqual('1');
  });
});
