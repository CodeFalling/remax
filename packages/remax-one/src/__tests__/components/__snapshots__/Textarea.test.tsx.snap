// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea render correctly 1`] = `
<textarea
  className="class"
  controlled={false}
  onBlur={[Function]}
  onConfirm={[Function]}
  onFocus={[Function]}
  onInput={[Function]}
  toutiao-selection-end={999}
  toutiao-selection-start={999}
  wechat-fixed={false}
  wechat-selection-end={999}
  wechat-selection-start={999}
/>
`;

exports[`Textarea should have maxlength when PLATFORM is wechat or toutiao 1`] = `
<textarea
  className="class"
  controlled={false}
  maxLength={140}
  onBlur={[Function]}
  onConfirm={[Function]}
  onFocus={[Function]}
  onInput={[Function]}
  toutiao-selection-end={999}
  toutiao-selection-start={999}
  wechat-fixed={false}
  wechat-selection-end={999}
  wechat-selection-start={999}
/>
`;
